hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSImageFormatProtobuf.java

```diff
@@ -294,6 +368,19 @@ public int compare(FileSummary.Section s1, FileSummary.Section s2) {
        * a particular step to be started for once.
        */
       Step currentStep = null;
+      boolean loadInParallel =
+          conf.getBoolean(DFSConfigKeys.DFS_IMAGE_PARALLEL_LOAD_KEY,
+              DFSConfigKeys.DFS_IMAGE_PARALLEL_LOAD_DEFAULT);
+      // TODO - check for compression and if enabled disable parallel
+
+      ExecutorService executorService = null;
+      ArrayList<FileSummary.Section> subSections =
+          getAndRemoveSubSections(sections);
+      if (loadInParallel) {
+        executorService = Executors.newFixedThreadPool(
+            conf.getInt(DFSConfigKeys.DFS_IMAGE_PARALLEL_THREADS_KEY,
+                DFSConfigKeys.DFS_IMAGE_PARALLEL_THREADS_DEFAULT));
```

Comment:
nice to have: log an info message stating that parallel image loading is enabled and the number of threads used.

Reply:
I added a log message here, and validated the thread count setting is not less than 1, otherwise I reset it to the default (4). I also pulled the code to do this check and to create the executor into a private method to reduce the noise in the already too long loadInternal method.